File: 卡码/代码随想录/Go/回溯/05-棋盘问题/LeetCode_0051.go
// 51. N 皇后
package main

import "strings"

func solveNQueens(n int) [][]string {
	res:=make([][]string,0)
	chess:=make([][]string,n)
	var backtracking func(row int)
	var isValid func(row,col int) bool
	isValid=func(row,col int) bool{
		for i:=0;i<row;i++ {
			if chess[i][col]=="Q" {
				return false
			}
		}
		for i,j:=row-1,col-1;i>=0&&j>=0;i,j=i-1,j-1 {
			if chess[i][j]=="Q" {
				return false
			}
		}
		for i,j:=row-1,col+1;i>=0&&j<n;i,j=i-1,j+1 {
			if chess[i][j]=="Q" {
				return false
			}
		}
		return true
	}
	backtracking=func(row int){
		if row==n {
			tmp:=make([]string,n)
			for i,str:=range chess {
				tmp[i]=strings.Join(str,"")
			}
			res=append(res,tmp)
			return
		}
		for col:=0;col<n;col++ {
			if isValid(row,col) {
				chess[row][col]="Q"
				backtracking(row+1)
				chess[row][col]="."
			}
		}
	}
	for i:=0;i<n;i++ {
		chess[i]=make([]string,n)
		for j:=0;j<n;j++ {
			chess[i][j]="."
		}
	}
	backtracking(0)
	return res
}

File: 卡码/代码随想录/Python/回溯/05-棋盘问题/LeetCode_0051.py
#51. N 皇后
from typing import *

class Solution:
  def __init__(self):
    self.chess:List[List[str]]=[]
    self.res:List[List[str]]=[]
  def isValid(self,row:int,col:int,n:int)->bool:
    for i in range(row):
      if self.chess[i][col]=='Q':
        return False
    i:int=row-1
    j:int=col-1
    while i>=0 and j>=0:
      if self.chess[i][j]=='Q':
        return False
      i,j=i-1,j-1
    i,j=row-1,col+1
    while i>=0 and j<n:
      if self.chess[i][j]=='Q':
        return False
      i,j=i-1,j+1
    return True
  def backtracking(self,row:int,n:int):
    if row==n:
      tmp:List[str]=[]
      for i in range(len(self.chess)):
        tmp.append("".join(self.chess[i]))
      self.res.append(tmp[:])
    for col in range(n):
      if self.isValid(row,col,n):
        self.chess[row][col]='Q'
        self.backtracking(row+1,n)
        self.chess[row][col]='.'
  def solveNQueens(self, n: int) -> List[List[str]]:
    self.chess=[['.']*n for _ in range(n)]
    self.backtracking(0,n)
    return self.res